---
title: "BEME: Microbiome data analysis"
author: "MAST190126 assignment 2 prepared by Kim-Anh LÃª Cao"
date: '\today'
#date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  pdf_document:
    
    toc: true
    toc_depth: 3
    number_sections: true
    citation_package: biblatex
  highlight: zenburn
bibliography: bibliography.bib
#csl: biomed-central.csl
csl: clinical-pharmacology-and-therapeutics.csl
header-includes:
- \usepackage{float}
- \floatplacement{figure}{H}
- \usepackage{fancyhdr}
- \usepackage{xcolor, hyperref}
- \usepackage{lipsum}
- \setlength{\headheight}{28pt}
- \setlength{\footskip}{25pt}
- \pagestyle{fancy}
- \renewcommand{\headrulewidth}{0.5pt}
- \renewcommand{\footrulewidth}{0.5pt}
- \rhead{\thepage}
- \hypersetup{colorlinks   = true, linkcolor=blue, urlcolor  = blue}
- \fancypagestyle{plain}{\pagestyle{fancy}}
- \cfoot{\scriptsize Melbourne Integrative Genomics, School of Mathematics and Statistics | The University of Melbourne, VIC 3010 \\ \url{http://mixomics.org/} | \url{http://lecao-lab.science.unimelb.edu.au/}}
editor_options: 
  chunk_output_type: console
---

```{r global_options, include=FALSE}
library(knitr)
# global options
knitr::opts_chunk$set(dpi = 100, echo=TRUE, warning=FALSE, message=FALSE, eval = TRUE,
                      fig.show=TRUE, fig.width= 7,fig.height= 6,fig.align='center', out.width = '50%', 
                      fig.path= 'Figures/')

# The libraries to load
library(mixOmics)
library(kableExtra)
library(MultiAssayExperiment)

# to increase mem size if needed
file.create(".Renviron")
cat("R_MAX_VSIZE=100Gb", file = ".Renviron")
```


Questions to Al:
- what is the normalisation applied to the RNA-seq data? 
  *noteFromAl*: These are scran-normalised data.
- what is the (most advised) function to filter only the most variable genes (RNA-seq)?
  *noteFromAl*: see below.
```{r}
## assumes majority of genes are non-DE
get_hvgs_rna <- function(log_counts = assay(gastru.mae, "rna"),
                         n_genes = 2000, ## No. of genes
                         use_bio=TRUE, ## choose based on biological variance or total variance?
                         do_plot = FALSE ## plot mean-variance dependency pre and post decomposition?
                         ){ 
  if (is())
  require(scran)
  fit <- trendVar(x = log_counts)
  decomposed_var <- decomposeVar(x = log_counts, fit = fit, use.spikes=FALSE)
  
  if (do_plot) {
    require(ggplot2)
    ## total variance
    p_aft <- ggplot(as.data.frame(decomposed_var)) + geom_point(aes(x=mean, y=total), alpha=0.4) +
      labs(x="mean log expression", y = "total variance")
    print(p_aft)
    ## decomposed bio
    p_bef <- ggplot(as.data.frame(decomposed_var)) + geom_point(aes(x=mean, y=bio), alpha=0.4) +
      labs(x="mean log expression", y = "decomposed biological variance")
    print(p_bef)
  }
  
  if (use_bio) {
    decomposed_var <- decomposed_var[order(-decomposed_var$bio),]
  } else {
     decomposed_var <- decomposed_var[order(-decomposed_var$total),]
  }
  hvgs <- rownames(decomposed_var)[1:n_genes]
  
  
  return(log_counts[hvgs,])
}
```

```{r}
## assumes majority of genes are non-DE
get_hvgs_sce <- function(sce = NULL,
                         n_genes = NULL, ## No. of genes, NULL for all genes ordered by variation
                         use_bio=TRUE ## choose based on biological variance or total variance?
                         ){ 
  ## -- decompose variance
  require(scran)
  fit <- trendVar(x = sce, use.spikes=FALSE)
  sce <- decomposeVar(x = sce, fit = fit, use.spikes=FALSE)
  ## -- order genes based on variance
  gene_order <- ifelse(use_bio, order(-sce$bio), order(-sce$total))
  
  ## -- subset genes (if asked)
  if(is.null(n_genes)) {
    n_genes <- dim(sce)[1]
  }
  
  keep_genes <- gene_order[seq_len(n_genes)]
  
  ## -- subset sce
  sce <- sce[keep_genes,]
  return(sce)
}
```

- what is the (most advised) function to filter only the top  methylation markers?
  *noteFromAl*: They're now ordered based on increasing biological variation.
  
- the PCA on the methylation marker takes a while to run. Any recommendation? (in general, for methylation data, see Q below)
  *noteFromAl*: I would not recommend PCA on methylation only. let me know if you want me to run the MDS
  
- what is the methylation data set you would recommend for the students to analyse (can you rank them from most interesting to less interesting). Similarly for the chromatin accessibility. I can ask them to test various combination of 3-omics data sets integration.
   *noteFromAl*: Biological insight into these cells is needed to answer this question. Off the top of my head, regulatroy regions such as P300, CTCF and DHS sites are quite interesting which I just processed and added to MAE.


# Load data
```{r}
# gastru.mae <- readRDS('/Users/alabadi/Projects/analysis/R/multiOmics/scNMT_seq_gatrulation/output/scnmtseq_gastrulation_mae.rds') 
gastru.mae <- readRDS('/Users/alabadi/Projects/analysis/R/multiOmics/scNMT_seq_gatrulation/output/scnmtseq_gastrulation_mae_826-cells_orderedFeatures.rds')  # MAE experiment
gastru.mae

# the meta data
colData(gastru.mae)

# meta data for 'rna' data set
summary(as.factor(colData(gastru.mae[,,'rna'])$lineage10x_2))
summary(as.factor(colData(gastru.mae[,,'rna'])$stage))
summary(as.factor(colData(gastru.mae[,,'rna'])$stage_lineage)) 

# ~ warning: 2 cells have no lineage. The NA may have to be coded differently in the mixOmics functions
#noteFromAl: Some of the cells are still in early days and, naturally, have not yet taken a lineage (hence NA is more of a Not-Applicable than Unknown) while all will have stages. That's why stage_lineage is a more specific phenotype to use. But I understand that the code must be flexible to support NA values in `group`. Let me know on Slack if you want me 
```

# Extract one dataset: transcriptomics
```{r}
X <- assay(gastru.mae[,,'rna'])
metadata.X <- colData(gastru.mae[,,'rna'])
# assign the NA lineage to Unknown
metadata.X$lineage10x_2[is.na(metadata.X$lineage10x_2)] = 'Unknown'
#ifelse(is.na(metadata.X$lineage10x_2), 'Unknown', metadata.X$lineage10x_2) 
summary(as.factor(metadata.X$lineage10x_2))

dim(X)

# transpose
X <- t(X)
dim(X)

# filter genes to keep highly variable
# across all samples
var.X = sort(apply(X, 2, var), decreasing = TRUE)
plot(var.X, type = 'h')
keep.X = names(var.X[1:10000])
length(keep.X)

X = X[, keep.X]
dim(X)
X[1:5,1:5]


library(mixOmics)
pca.X = pca(X, ncomp = 5)
plot(pca.X)
plotIndiv(pca.X, group = metadata.X$stage, legend = TRUE, ind.names =FALSE)
plotIndiv(pca.X, group = metadata.X$lineage10x_2, pch = 16, legend = TRUE, ind.names =FALSE)

plotIndiv(pca.X, group = metadata.X$lineage10x_2, pch = 16, legend = TRUE, ind.names =FALSE, comp = c(1,3))
```


# Extract one dataset: methylation
```{r}
X <- assay(gastru.mae[,,'met_genebody'])
metadata.X <- colData(gastru.mae[,,'met_genebody'])
# assign the NA lineage to Unknown
metadata.X$lineage10x_2[is.na(metadata.X$lineage10x_2)] = 'Unknown'
#ifelse(is.na(metadata.X$lineage10x_2), 'Unknown', metadata.X$lineage10x_2) 
summary(as.factor(metadata.X$lineage10x_2))

dim(X)

# transpose
X <- t(X)
dim(X)

# filter genes to keep highly variable
# across all samples
# ! there are many missing values, include na.rm = TRUE in var calculation
var.X = sort(apply(X, 2, var, na.rm = TRUE), decreasing = TRUE)
plot(var.X, type = 'h')
keep.X = names(var.X[1:5000])   # for methylation markers, lots of NA will make NIPALS run, i.e. slow!
length(keep.X)

X = X[, keep.X] 
dim(X)
X[1:5,1:5]


library(mixOmics)
pca.X = pca(X, ncomp = 5)  # long to run, any trick here?
plot(pca.X)
plotIndiv(pca.X, group = metadata.X$stage, legend = TRUE, ind.names =FALSE)
plotIndiv(pca.X, group = metadata.X$lineage10x_2, pch = 16, legend = TRUE, ind.names =FALSE)

plotIndiv(pca.X, group = metadata.X$lineage10x_2, pch = 16, legend = TRUE, ind.names =FALSE, comp = c(1,3))
```
