---
title: "scNMT-seq multi-omics analysis"
output: html_document
params:
  on_mac: !r Sys.info()["user"] == "alabadi"
  on_hpc: !r Sys.info()["user"] != "alabadi"
  cache: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = params$cache)
```

This analysis build on: https://github.com/BIRSBiointegration/Hackathon/tree/master/scNMT-seq

# TO Do

* commit changes so tracking is easier
* A way/function to get consistent sample names for assays when creating the diablo's X for rna-promoter-p300-diablo.Rmd and others >> how did Lucy do it? Or how did we do it in the 'explore' file?

# Directions

- Get DIABLO LVs with stages / lineages and different design matrices >> either per-block or (weighted) consensus variates >> investiagte what would make sense.
- Use metrics mentioned in https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9 to evaluate the resulting 'clusters'
- Run a K-means/median clustering using groud truth number of clusters for MOFA?
- Look at LIGER and Harmony as well and see if you can come up with a crieria to compare


i) Compare ARI and other benchmarking metrics in https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9 using KNN with stages/lineages as 

# RNA

* [RNA gene filtering based on expression](notebook/filter-rna-data.html)
* [Comparison of normalisation methods for scRNAseq data](notebook/compare-rna-normalisation.html)
* [Add pearson residuals to MAE and rank rna features based on variance](notebook/update-rna-in-mae.html)
<!-- * [Predict unknown lineages using a PLSDA model with RNA data]() -->

# Imputation

* [imputation benchmarkig using means vs NIPALS](notebook/imputation.html)
* [impute MAE](notebook/imputation.html)

# Integration

* [DIABLO - RNA - promoters - enhancers (markers + error rates)](notebook/)
* [MOFA: RNA - promoters - enhancers (markerss)](notebook/)
* [k-means silhouette/purity using block.spls RNA ~ promoters + enhancers vs MOFA variates ](notebook/)