---
title: "block.splsda excluidng rna using annot as Y"
author: "Al J Abadi"
params:
  on_hpc: !r Sys.info()['user'] != 'alabadi'
  on_mac: !r Sys.info()['user'] == 'alabadi'
  rerun: false
output:
  html_document:
    toc: true
---

```{r, warning=FALSE, message=FALSE}
library(MultiAssayExperiment)
library(mixOmics)
library(here)
```


```{r, echo=FALSE}
knitr::opts_chunk$set(eval = TRUE, cache = !FALSE, warning = FALSE, out.width = "100%")
```


```{r}
source(here("src/source-utils.R"))
```

## Data

```{r}
gastru.mae <- readRDS(here('output/scnmtseq_gastrulation_mae_826-cells_orderedFeatures.rds'))  # MAE experiment
gastru.mae <- gastru.mae[,!is.na(gastru.mae$lineage10x_2),]
```


check column names are consistent:

```{r}
all_identical(colnames(gastru.mae))
```

## designate assays

```{r}
study_assays <- c("met_promoter", "acc_promoter")
# study_assays <- c("rna", "met_genebody", "met_promoter", "met_cgi",  "acc_promoter", "acc_cgi",  "acc_p300", "acc_CTCF", "acc_DHS")
```


```{r}
study_mae <- gastru.mae[,,study_assays]
# study_mae <- study_mae[,(study_mae$day %in%  c(4.5, 7.5)) 
#                        # & (study_mae$lineage10x_2 %in% c("Epiblast", "Mesoderm", "Endoderm", "Ectoderm"))
#                       ,]
coldata <- data.frame(colData(study_mae))
# coldata$lineage10x_2[is.na(coldata$lineage10x_2)] <- "Unknown"
```

```{r}
lapply(named_list(unique(coldata$stage)), function(stage){
  table(coldata[coldata$stage == stage,]$lineage10x_2)
})
# $E5.5
#          Epiblast Visceral_endoderm 
#                84                24 
# $E4.5
#           Epiblast Primitive_endoderm 
#                 60                 43 
# $E6.5
#          Epiblast      ExE_ectoderm          Mesoderm  Primitive_Streak Visceral_endoderm 
#               146                 8                28                43                45 
# $E7.5
#         Ectoderm         Endoderm         Epiblast         Mesoderm Primitive_Streak 
#               43               81               44              141               33 
```


```{r}
dput(unique(coldata[coldata$stage == 'E7.5',]$lineage10x_2))
mae_E7.5 <- gastru.mae[,(gastru.mae$stage == 'E7.5') & (gastru.mae$lineage10x_2 %in% c("Endoderm", "Epiblast", "Mesoderm", "Ectoderm"))]
```

```{r}
mae <- mae_E7.5
top_n <- 500
assay_i <- 'met_promoter'
dist <- 'max.dist'

X <- t(assay(mae, assay_i))
if (!is.null(top_n)) {
  X <- X[,seq_len(top_n)]
}
Y <- mae$lineage10x_2
plsda.res <- splsda(X = X, Y = Y, ncomp = 2, keepX = c(50,50), max.iter = 500)
folds <- min(10, min(table(mae$lineage10x_2)))
stopifnot(folds>5)
perf.plsda.res <- perf(plsda.res, folds = folds, nrepeat = 3, dist = 'max.dist', cpus = 4)
accur <- perf.plsda.res$error.rate.class[[dist]]
#   Ectoderm   Endoderm   Epiblast   Mesoderm 
# 1.00000000 0.98353909 0.73484848 0.04018913 
top_features <- perf.plsda.res$features$stable$comp1
top_features <- top_features[top_features>=0.5]

```

GO term enrichment for most features
```{r}
feas_genes <- rep(0, length(rownames(assay(mae, assay_i))))
names(feas_genes) <- rownames(assay(mae, assay_i))
feas_genes[names(top_features)] <- 1
GOdata <- GOget(feas_genes = feas_genes, ID = 'ensembl')
goEnrich.res <- goEnrich(GOdata = GOdata)
GOplot.res <- GOplot(goEnrichment = goEnrich.res, trms = c('metabol', 'catabol'))
```


```{r}
#' Plot boxplots of mean rate/expression of features across cells
#'
#' 
#' @param X An array of data with (n x p)
#' @param coldata colData(mae) with matching cell names
#' @param selected_features A named vector of loadings values
#' @param pheno Phenotype column name in coldata
#' @param type choice(s) of c('box', 'violin')
#' @param x_lab x label
#' @param y_lab y label
#' @param weights If '+'/'-', only '+'/'-' loadings are considered
#' @param show_enrichment Whether to plot the global average. Currently not used.
#' @param title Plot title
#'
#' @return
#' @export
#'
#' @examples
boxplot_means <- function(X, coldata, selected_features, pheno = 'lineage10x_2', x_lab = 'Phenotype', type = c('box', 'violin'),  y_lab = 'Rate', weights = "+", show_enrichment = FALSE, title = NULL ) {
  type <- match.arg(type, several.ok = TRUE)
  library(data.table)
  coldata_dt <- data.table(coldata, keep.rownames = "cell", check.names = FALSE, stringsAsFactors = FALSE)
  if (weights == "+") {
    selected_features <- selected_features[selected_features > 0]
  } else if (weights == "-") {
    selected_features <- selected_features[selected_features < 0]
  }
  if (length(selected_features) < 1) {
    message("No features with requested weights")
    return(NULL)
  }
    

  dt_background <- data.table(X, keep.rownames = "cell", check.names = FALSE, stringsAsFactors = FALSE)
  dt_background <- melt.data.table(dt_background, variable.name = "feature", value.name = "rate")
  dt_background <- data.table::merge.data.table(dt_background, coldata_dt)

  dt <- dt_background[feature %in% names(selected_features)]

  df_background <- data.frame(dt_background[,.(mean_rate = mean(rate, na.rm=TRUE)), by = c(pheno, "feature")])
  df <- data.frame(dt[,.(mean_rate = mean(rate, na.rm=TRUE)), by = c(pheno, "feature")])
  p <- ggplot(df, aes(x = !!as.name(pheno), y = mean_rate))
  # p <- p + geom_boxplot(data = df_background, aes(x = !!as.name(pheno), y = mean_rate), fill = "grey75", alpha = 0.75, outlier.shape = NA, coef = 0)
  if (type == 'box' %in% type) {
    p <- p + geom_boxplot(aes(fill = !!as.name(pheno)), width = 0.3 ) 
  } else if (type == 'violin') {
    p <- p +  geom_violin(aes(fill = !!as.name(pheno)), width = 0.3)
  } else {
    p <- p + 
      geom_violin(aes(fill = !!as.name(pheno)), width = 0.35) +
      geom_boxplot(aes(fill = !!as.name(pheno)), width = 0.1 )
  }
  p <- p +
    labs(x = x_lab, y = y_lab, title = title)  +  theme_bw()  +
    geom_jitter(size = 1, col = "grey30", position = position_jitter(0.1))

  p <- p + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))
  
  list(plot=p, df=df)
}
```

```{r}
comp <- 1
selected_features <- selectVar(plsda.res, comp = comp)$value[,1][1:10]
names(selected_features) <- selectVar(plsda.res, comp = comp)$name[1:10]
boxplot_means(X = X, coldata = coldata, selected_features = selected_features, weights = '+')
negative <- boxplot_means(X = X, coldata = coldata, selected_features = selected_features, weights = '-')
```

Top features
```{r}
selectVar(plsda.res)
```


## run block.splsda

```{r}
X <- lapply(as.list(experiments(study_mae)), t)
X <- suffix_colnames(X, indices = seq_along(X)[-1])

# if (params$on_mac) {
#   X <- lapply(X,  subset_pn, p=2000)
# }
lapply(X, dim)
Y <- coldata$lineage10x_2
```


```{r, eval = params$rerun}
design <- create_design(mae = study_mae, off_diag = 0)

tune_diablo <- tune.block.splsda(X = X, Y = Y, test.keepX = lapply(X, function(x) c(5,10, 20, 50)), ncomp = 3, design = design, folds = 10, nrepeat = 3, near.zero.var = TRUE, cpus = 4)
keepX <- create_keepX(study_mae, keepX = c(50, 50))

block.splsda.res <- block.splsda(X = X, Y = Y, ncomp = 2, keepX = keepX)
```

```{r, eval = params$rerun}
saveRDS(block.splsda.res, file = here("output/block.splsda.annot-vs-epigenome.res.rds"))
```

```{r, eval = !params$rerun}
block.splsda.res <- readRDS(here("output/block.splsda.annot-vs-epigenome.res.rds"))
```


```{r}
# pdf("diablo-FirstAndLastStages-AllLineages.pdf", width = 12, height = 8)
plotIndiv(block.splsda.res, pch=factor(coldata$lineage10x_2), group = study_mae$stage, legend = TRUE, comp = 1:2, blocks = seq_along(X), legend.title = "Stage", legend.title.pch = "Lineage", cex = 1, size.subtitle = 10)
# dev.off()
# plotIndiv(block.splsda.res, pch=16, group = coldata$lineage10x_2, legend = TRUE, comp = 1:2, blocks = seq_along(X), legend.title = "Lineage")
```

Methylation in promoters and accessibility in p300 enhancers seem to correlate with embryonic days:

```{r}
library(data.table)
coldata_dt <- data.table(coldata, keep.rownames = "cell", check.names = FALSE, stringsAsFactors = FALSE)
selected_features <- selectVar(block.splsda.res)[seq_along(X)]
```

### mean rate for selected features by stage

```{r}
boxplot_means <- function(omic, x_lab, y_lab, weights = "+", show_enrichment = FALSE, title = NULL ) {
  
  keep_features <- selected_features[[omic]]$name
  
  if (weights == "+") {
    keep_features <- keep_features[selected_features[[omic]]$value$value.var > 0]
  } else if (weights == "-") {
    keep_features <- keep_features[selected_features[[omic]]$value$value.var < 0]
  }
  if (length(keep_features) < 1)
    stop("No features with requested weights")
  
  dt_background <- data.table(X[[omic]], keep.rownames = "cell", check.names = FALSE, stringsAsFactors = FALSE)
  dt_background <- melt.data.table(dt_background, variable.name = "feature", value.name = "rate")
  dt_background <- data.table::merge.data.table(dt_background, coldata_dt)
  
  dt <- dt_background[feature %in% keep_features]
  
  df_background <- data.frame(dt_background[,.(mean_rate = mean(rate, na.rm=TRUE)), by = c("lineage10x_2", "feature")])
  df <- data.frame(dt[,.(mean_rate = mean(rate, na.rm=TRUE)), by = c("lineage10x_2", "feature")])
  p <- ggplot(df, aes(x = lineage10x_2, y = mean_rate))  
  p <- p + geom_boxplot(data = df_background, aes(x = lineage10x_2, y = mean_rate), fill = "grey75", alpha = 0.75, outlier.shape = NA, coef = 0)
  
  p <- p + geom_violin(aes(fill = lineage10x_2))  +  geom_boxplot(aes(fill = lineage10x_2), width = 0.1, alpha = 0.4 )  + 
    labs(x = x_lab, y = y_lab, title = title)  +  theme_bw()  + 
    geom_jitter(size = 1, col = "grey30", position = position_jitter(0.3))
  
  
  list(plot=p, df=df)
}
```

For Pathway Analysis

```{r}
library(biomaRt)
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
```

```{r}
get_bm <- function(genes) {
  getBM(
  filters = "ensembl_gene_id",
  attributes = c("ensembl_gene_id", "entrezgene_id"),
  values = genes,
  mart = mouse
)
}
```
```{r}
# enrich_sig_pathway <- function(block) {
# c
# }
```

#### rna

```{r}
boxplot_means(omic = "rna", x_lab = "Embryonic Stage", y_lab = "mean gene expression across cells", weights = "both", title = "All Features")
boxplot_means(omic = "rna", x_lab = "Embryonic Stage", y_lab = "mean gene expression across cells", weights = "+", title = "posititive weights")
```

```{r}
rna_negative <- boxplot_means(omic = "rna", x_lab = "Embryonic Stage", y_lab = "mean gene expression across cells", weights = "-", title = "negative weights")
rna_negative$plot
```


```{r}
rna_reactome <- enrich_sig_pathway(block = "rna")
summary(rna_reactome)
```

#### met_promoter

No negative weights

```{r}
met_promoter <- boxplot_means(omic = "met_promoter", x_lab = "Embryonic Stage", y_lab = "mean promoter methylation", weights = "+")
met_promoter$plot
```

```{r}
genes <- stringr::str_extract(string = selected_features$met_promoter$name, pattern = "^ENSMUSG[0-9]*")
genes_BM <- get_bm(genes = genes)
library(ReactomePA)
rna_reactome <- enrichPathway(genes_BM$entrezgene_id, organism = "mouse", readable = TRUE)
summary(rna_reactome)


library(ReactomePA)
met_promoter_reactome <- enrichPathway(genes_BM$entrezgene_id, organism = "mouse", readable = TRUE)
summary(met_promoter_reactome)
```

#### met_genebody

No negative weights

```{r}
met_genebody <- boxplot_means(omic = "met_genebody", x_lab = "Embryonic Stage", y_lab = "mean genebody methylation")
met_genebody$plot
```
```{r}
met_genebody$df
```
```{r}
genes <- stringr::str_extract(string = met_genebody$df$feature, pattern = "^ENSMUSG[0-9]*")
library(biomaRt)
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
genes_BM <- getBM(
  filters = "ensembl_gene_id",
  attributes = c("ensembl_gene_id", "entrezgene_id"),
  values = genes,
  mart = mouse
)
library(ReactomePA)
genes_reactome <- enrichPathway(genes_BM$entrezgene_id, organism = "mouse", readable = TRUE)
summary(genes_reactome)
```

```{r}
barplot(genes_reacome)
emapplot(genes_reacome)
cnetplot(genes_reacome)
```


#### acc_p300

```{r}
boxplot_means(omic = "acc_p300", x_lab = "Embryonic Stage", y_lab = "mean enhancer accessibility")
```

#### acc_DHS

```{r}
boxplot_means(omic = "acc_p300", x_lab = "Embryonic Stage", y_lab = "mean gene expression across cells")
```

```{r}
# plotVar(block.spls.res, var.names = FALSE, pch = c(16:21), cutoff = 0.5, legend = TRUE)
```

```{r}
# corMat <- network(block.spls.res, save = "pdf", name.save = "block.spls.network", cutoff = 0.81)
```




```{r}
# library(knitr)
# include_graphics("block.spls.network.pdf")
```

## run block.spls

```{r, eval = params$rerun}
Y_both <- cbind(gastru.mae$day, unmap(gastru.mae$lineage10x_2)) %>% set_colnames(c("day", unique(gastru.mae$lineage10x_2))) %>% set_rownames(rownames(X[[1]]))
block.spls.res <- block.spls(X = X, Y = Y_both, ncomp = 2, keepX = keepX, scale = FALSE)
saveRDS(block.spls.res, file = here("output","block.spls.res-Y-both-stage-lineage.rds"))
```
```{r, eval = !params$rerun}
block.spls.res <- readRDS(here("output","block.spls.res-Y-both-stage-lineage.rds"))
```

```{r}
plotIndiv(block.spls.res, pch=16, group = study_mae$day, legend = TRUE, comp = 1:2, blocks = seq_along(X), legend.title = "Stage")
plotIndiv(block.spls.res, pch=16, group = study_mae$lineage10x_2, legend = TRUE, comp = 1:2, blocks = seq_along(X), legend.title = "Lineage")
```

