---
title: "block.spls using rna, promoter, and p300 enhancer sites"
author: "Al J Abadi"
params:
  on_hpc: !r Sys.info()['user'] != 'alabadi'
  on_mac: !r Sys.info()['user'] == 'alabadi'
output:
  html_document:
    toc: true
---

```{r, warning=FALSE, message=FALSE}
library(MultiAssayExperiment)
library(mixOmics)
library(here)
```


```{r, echo=FALSE}
knitr::opts_chunk$set(eval = TRUE, cache = !FALSE, warning = FALSE, out.width = "100%")
```


```{r}
source(here("src/source-utils.R"))
```

# Data

```{r}
gastru.mae <- readRDS(here('output/scnmtseq_gastrulation_mae_826-cells_orderedFeatures.rds'))  # MAE experiment
gastru.mae <- gastru.mae[,!is.na(gastru.mae$lineage10x_2),]
```


check column names are consistent:

```{r}
all_identical(colnames(gastru.mae))
```

## designate assays

```{r}
study_assays <- c("rna", "met_promoter", "acc_p300", "met_genebody", "acc_DHS")
# study_assays <- c("rna", "met_genebody", "met_promoter", "met_cgi",  "acc_promoter", "acc_cgi",  "acc_p300", "acc_CTCF", "acc_DHS")
```


```{r}
study_mae <- gastru.mae[,,study_assays]
coldata <- data.frame(colData(study_mae))
coldata$lineage10x_2[is.na(coldata$lineage10x_2)] <- "Unknown"
```

## run block.spls

```{r}
X <- lapply(as.list(experiments(study_mae)), t)
X <- suffix_colnames(X, indices = seq_along(X)[-1])

# if (params$on_mac) {
#   X <- lapply(X,  subset_pn, p=2000)
# }
lapply(X, dim)
Y <- matrix(study_mae$day, dimnames = list(rownames(X[[1]]), "Day"))
```


```{r, eval=params$on_hpc}
design <- create_design(mae = study_mae, off_diag = 1, rna_only = TRUE)
keepX <- create_keepX(study_mae, keepX = c(50, 50))

block.spls.res <- block.spls(X = X, Y = Y, ncomp = 2, keepX = keepX, scale = TRUE)
saveRDS(block.spls.res, file = here("output/block.spls.res.rds"))
```

```{r}
block.spls.res <- readRDS(here("output/block.spls.res.rds"))
```


```{r}
plotIndiv(block.spls.res, pch=16, group = study_mae$stage, legend = TRUE, comp = 1:2, blocks = seq_along(X), legend.title = "Stage")
plotIndiv(block.spls.res, pch=16, group = coldata$lineage10x_2, legend = TRUE, comp = 1:2, blocks = seq_along(X), legend.title = "Lineage")
```

Methylation in promoters and accessibility in p300 enhancers seem to correlate with embryonic days:

```{r}
library(data.table)
coldata_dt <- data.table(coldata, keep.rownames = "cell", check.names = FALSE, stringsAsFactors = FALSE)
selected_features <- selectVar(block.spls.res)[seq_along(X)]
```

### mean rate for selected features by stage

```{r}
boxplot_means <- function(omic, x_lab, y_lab) {
  dt <- data.table(X[[omic]][,selected_features[[omic]]$name], keep.rownames = "cell", check.names = FALSE, stringsAsFactors = FALSE)
  dt <- melt.data.table(dt, variable.name = "feature", value.name = "rate")
  dt <- data.table::merge.data.table(dt, coldata_dt)
  df <- data.frame(dt[,.(mean_rate = mean(rate, na.rm=TRUE)), by = c("day", "feature")])
  df$stage <- paste0("E", df$day)
  ggplot(df)  +  geom_violin(aes(x = stage, y = mean_rate, fill = stage))  +  geom_boxplot(aes(x = stage, y = mean_rate, fill = stage), width = 0.1, alpha = 0.4 )  + 
    labs(x = x_lab, y = y_lab)  +  theme_bw()
}
```

#### rna


```{r}
boxplot_means(omic = "rna", x_lab = "Embryonic Stage", y_lab = "mean gene expression across cells")
```

#### met_promoter

```{r}
boxplot_means(omic = "met_promoter", x_lab = "Embryonic Stage", y_lab = "mean promoter methylation")
```

#### met_genebody

```{r}
boxplot_means(omic = "met_genebody", x_lab = "Embryonic Stage", y_lab = "mean genebody methylation")
```

#### acc_p300

```{r}
boxplot_means(omic = "acc_p300", x_lab = "Embryonic Stage", y_lab = "mean enhancer accessibility")
```

#### acc_DHS

```{r}
boxplot_means(omic = "acc_p300", x_lab = "Embryonic Stage", y_lab = "mean DHS accessibility")
```

```{r}
# plotVar(block.spls.res, var.names = FALSE, pch = c(16:21), cutoff = 0.5, legend = TRUE)
```

```{r}
# corMat <- network(block.spls.res, save = "pdf", name.save = "block.spls.network", cutoff = 0.81)
```

```{r}
# library(knitr)
# include_graphics("block.spls.network.pdf")
```