---
title: "scnmtseq-gastrulation"
output: html_document
params:
  on_hpc: !r Sys.info()["user"] != "alabadi"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

## I/O

```{r io}
#########
## I/O ##
#########
io$data <- io$out <- io <- list()
## Rproject/slurm file home - must be FULL PATH
io$home <- ifelse(Sys.info()["user"] == "alabadi",
       "/Users/alabadi/Projects/analysis/R/multiOmics/scNMT_seq_gatrulation",
       "/data/cephfs/punim0613/AL/Projects/analysis/R/multiOmics/scNMT_seq_gatrulation")
## outputs
io$outdir <- file.path(io$home, "output")
```

## utils

```{r utls}
## ----------- source utils ----------- 
## list all R files in utils folder recursively
utls <- list.files(file.path(io$home, "src/utils"), pattern = ".R", full.names = TRUE, recursive = TRUE)
## source all
invisible(sapply(utls, source))
```

```{r, eval=params$on_hpc}
## install requirted packages if not installed
installer(pkgs = c("data.table", "magrittr", "scater", "purrr", "parellel"))
```

# Packages

```{r}
suppressMessages(library(data.table))
suppressMessages(library(scater))
suppressMessages(library(purrr))
suppressMessages(library(parallel))
suppressMessages(library(magrittr))
```

# Initialise

## options

```{r}
#############
## Options ##
#############
opts <- list()
rtimes <- list() ## record run times

opts$met.annos <- c(
  Promoters = "prom_2000_2000",
  Genebody = "genebody",
  CGI = "CGI",
  p300 = "ESC_p300",
  CTCF = "ESC_CTCF",  ##CTCF binding sites.  The primary role of CTCF transcription factors is thought to be in regulating the 3D structure of chromatin.
  DHS = "ESC_DHS" ## In these specific regions of the genome, chromatin has lost its condensed structure, exposing the DNA and making it accessible.
  
)

opts$acc.annos <- opts$met.annos
```

## data directories

```{r}
## files/directories from data-alias that we'll need
io$data$base   <- file.path(io$home, "data/gastrulation") ## data dir
io$data$sample.metadata <- file.path(io$data$base,"sample_metadata.txt")
# io$annos.dir  <- file.path(io$base_dir, "features/filt")
io$data$rna.file   <- file.path(io$data$base, "rna/parsed/SingleCellExperiment.rds")
io$data$met.dir   <- file.path(io$data$base, "met/parsed")
io$data$acc.dir   <- file.path(io$data$base, "acc/parsed")
```


```{r}
## input files
io$data$mets <- lapply(opts$met.annos, function(x) sprintf("%s/%s.tsv.gz", io$data$met.dir, x))
io$data$accs <- lapply(opts$acc.annos, function(x) sprintf("%s/%s.tsv.gz", io$data$acc.dir, x))
```

```{r cechk-files-exist}
## check that all files/dirs are valid
check_files_exist(io)
```

# Data

## metadata

```{r}
sample_metadata <- fread(io$data$sample.metadata)
# sample_metadata %>% head() %>% View()
sample_metadata <- sample_metadata %>% 
  .[pass_metQC==TRUE & pass_rnaQC==TRUE  & pass_accQC==TRUE ] %>% 
  .[,c("sample","id_met","id_acc","id_rna","stage","lineage10x_2")] %>%
  .[,stage_lineage:=as.factor(paste(stage,lineage10x_2,sep="_"))]

length(unique(sample_metadata$sample)) ## sample size
```



## epigenetic data

### methylation

```{r}
io$out$met_dt_list <- file.path(io$outdir, "met_dt_list_scaled_wt.rds")
```

##### calculate site stats

```{r read-met-dt, eval=params$on_hpc}
rtimes$met_dt_list <- system.time({ ## 48 sec
  
  met_dt_list <- mclapply(io$data$mets, function(tsv_name) {
    
    calc_site_stats(filePath = tsv_name,
                    sample_name = "id_met", ## name of sample to output
                    keep_samples = sample_metadata$id_met, ## pass QC samples
                    min_N = 3, ## min number of calls at site
                    min_cov = 400, ## min number of cells detecting
                    alpha = 0.1) ## for lower bound of CI
  }, mc.cores = parallel::detectCores())
  
})

saveRDS(met_dt_list, file = io$out$met_dt_list)
```

```{r}
# met_dt_list <- list()
# 
# for (tsv_i in seq_along(io$data$mets)) {
#   tsv_name <- names(io$data$mets)[tsv_i]
#   tsv_file <- io$data$mets[[tsv_i]]
#   cat(sprintf("Processing %s ...\n", tsv_name))
#   met_dt_list[[tsv_name]] <- tryCatch(calc_site_stats(filePath = tsv_file,
#                     sample_name = "id_met", ## name of sample to output
#                     keep_samples = sample_metadata$id_met, ## pass QC samples
#                     min_N = 3, ## min number of calls at site
#                     min_cov = 400, ## min number of cells detecting
#                     alpha = 0.1), error = function(e) e)
# }
```



### accessibility

```{r}
io$out$acc_dt_list <- file.path(io$outdir, "acc_dt_list-scaled_wt.rds")
```

##### calculate site stats

```{r read-acc-dt, eval=params$on_hpc}
rtimes$acc_dt_list <- system.time({ ## 67 sec
  
  acc_dt_list <- mclapply(io$data$accs, function(tsv_name) {
    
    calc_site_stats(filePath = tsv_name,
                    sample_name = "id_acc", ## name of sample to output
                    keep_samples = sample_metadata$id_acc, ## pass QC samples
                    min_N = 3, ## min number of calls at site
                    min_cov = 500, ## min number of cells detecting
                    alpha = 0.1) ## for lower bound of CI
  }, mc.cores = parallel::detectCores())
  
})

saveRDS(acc_dt_list, file = io$out$acc_dt_list)
```

### create data matrices

#### methylation

create wide matrices for MAE

```{r}
met_dt_list <- readRDS(io$out$met_dt_list)
```
```{r dcast-met}
## 8 sec
dcast_met <- dcast_dt_list(met_dt_list, row = "id", col = "id_met", value.var = "rate")
```

## get weights

```{r}
## 8 sec
dcast_met_scaled_wt <- dcast_dt_list(met_dt_list, row = "id", col = "id_met", value.var = "scaled_wt")
```


```{r}
lapply(dcast_met, dim)
```


```{r}
## quick check datasets to ensure features are properly ranked
mapply(x=dcast_met, y=met_dt_list, FUN = function(x,y) {
  ## is the first rowname in matrices the same as the max lci in data.table?
  identical(x %>% rownames() %>% .[1],
            y %>% .[which.max(.$lci)] %>% .$id)
})
```

#### accessibility

create wide matrices for MAE

```{r}
acc_dt_list <- readRDS(io$out$acc_dt_list)
```

```{r dcast-acc}
dcast_acc <- dcast_dt_list(acc_dt_list, row = "id", col = "id_acc", value.var = "rate")
```

## get weights

```{r}
dcast_acc_scaled_wt <- dcast_dt_list(acc_dt_list, row = "id", col = "id_acc", value.var = "scaled_wt")
```

```{r}
lapply(dcast_acc, dim)
```

```{r}
## quick check datasets to ensure features are properly ranked
mapply(x=dcast_acc, y=acc_dt_list, FUN = function(x,y) {
  ## is the first rowname in matrices the same as the max lci in data.table?
  identical(x %>% rownames() %>% .[1],
            y %>% .[which.max(.$lci)] %>% .$id)
})
```

## expression data

### load

<!-- Load RNA data -->
```{r read-sce}
# Load scater object
sce <- readRDS(io$data$rna.file)

# rna_passQC <- sample_metadata[pass_rnaQC==TRUE]$id_rna
# Filter cells
sce_matching <- sce[,colnames(sce) %in% sample_metadata$id_rna]
```

### change sample names and re-order

```{r}
## rename id_rna for sce
id_map <- sample_metadata[,c("sample", "id_rna")] %>% data.frame(row.names = 2)
colnames(sce_matching) <- id_map[colnames(sce_matching),]
```

Order cells based on epigenetic data

```{r}
all_identical(c(lapply(dcast_acc, colnames), lapply(dcast_met, colnames)))
# all_identical(c(lapply(dcast_acc, colnames), lapply(dcast_met, colnames), list(rna = colnames(sce_matching))))
```

```{r}
sce_matching <- sce_matching[, colnames(dcast_acc[[1]])]
```

### gene filtering and order

#### order genes based on variance

```{r}
sce_matching <- sce_matching[order(rowVars(logcounts(sce_matching)), decreasing = TRUE) ,]
```


#### histpgram of % of cells not expressing every given gene

```{r}
hist(rowData(sce_matching)$pct_dropout_by_counts, main = "Histogram of percent of cell dropouts for genes")
```

#### mean expression histogram

Look at mean expression histogram for all genes:
```{r}
hist_mean_exp <- function(sce, fill_col = "#3381A7") {
  rowdata <- as.data.frame(rowData(sce))
  ggplot(rowdata) + geom_histogram(aes(log10_mean_counts), bins = 50, fill = fill_col) + theme_bw()
}

hist_mean_exp(sce_matching)
```

#### table of gene dropout rates

```{r}
table_dropouts <- function(sce, pct_cells_expresing = c(0, 1, 5, 10)) {
  out <- data.frame(pct_cells_expressing = pct_cells_expresing, Number_of_genes = NA)
  out$Number_of_genes <- sapply(pct_cells_expresing, function(x){
    sum(rowData(sce)$pct_dropout_by_counts <= 100-x)
  })
  out
}

table_dropouts(sce_matching)

```


### genes to be expressed in at least 1% of cells

```{r}
sce_matching <- sce_matching[rowData(sce_matching)$pct_dropout_by_counts <= 99 ,]
dim(sce_matching)
```

### mean expression histogram after filtering

```{r}
hist_mean_exp(sce_matching)
```

```{r}
saveRDS(sce_matching, file = "easy-data/data/rna/parsed/SingleCellExperiment_matching-cells.rds")
```


# MultiAssayExperiment

## experimentlist

```{r build-exper-list}
library(MultiAssayExperiment)

exper.list <- ExperimentList(rna = logcounts(sce_matching),
                             
                             met_genebody = dcast_met$Genebody,
                             met_promoter = dcast_met$Promoters,
                             met_cgi = dcast_met$CGI,
                             met_p300 = dcast_met$p300,
                             met_CTCF = dcast_met$CTCF,
                             met_DHS = dcast_met$DHS,
                             
                             acc_genebody = dcast_acc$Genebody,
                             acc_promoter = dcast_acc$Promoters,
                             acc_cgi = dcast_acc$CGI,
                             acc_p300 = dcast_acc$p300,
                             acc_CTCF = dcast_acc$CTCF,
                             acc_DHS = dcast_acc$DHS,
                             
                             ## weights
                             wt_met_genebody = dcast_met_scaled_wt$Genebody,
                             wt_met_promoter = dcast_met_scaled_wt$Promoters,
                             wt_met_cgi = dcast_met_scaled_wt$CGI,
                             wt_met_p300 = dcast_met_scaled_wt$p300,
                             wt_met_CTCF = dcast_met_scaled_wt$CTCF,
                             wt_met_DHS = dcast_met_scaled_wt$DHS,

                             wt_acc_genebody = dcast_acc_scaled_wt$Genebody,
                             wt_acc_promoter = dcast_acc_scaled_wt$Promoters,
                             wt_acc_cgi = dcast_acc_scaled_wt$CGI,
                             wt_acc_p300 = dcast_acc_scaled_wt$p300,
                             wt_acc_CTCF = dcast_acc_scaled_wt$CTCF,
                             wt_acc_DHS = dcast_acc_scaled_wt$DHS
)
## check that all column names are identical
all_identical(lapply(exper.list, colnames))
```

## coldata

```{r}
coldata <- sample_metadata %>% data.frame(row.names = TRUE) %>% .[,c("stage", "lineage10x_2", 
"stage_lineage")] %>% DataFrame()
coldata <- coldata[colnames(sce_matching),]
coldata$day <- as.numeric(gsub("E", "", coldata$stage))
```

```{r}
all(coldata@rownames %in% colnames(sce_matching))
identical(coldata@rownames, colnames(dcast_acc[[1]]))
```

```{r}
coldata[is.na(coldata$lineage10x_2),]
```

## build MAE

```{r build-mae}
scnmtseq_gastrulation_mae <- MultiAssayExperiment(experiments = exper.list, colData = coldata)
```



```{r save-mae}
saveRDS(scnmtseq_gastrulation_mae, file= file.path(io$outdir, "scnmtseq_gastrulation_mae_826-cells_orderedFeatures.rds"))
```



