---
title: "scnmtseq-gastrulation"
output: html_document
params:
  on_mac: !r Sys.info()["user"] == "alabadi"
  on_hpc: !r Sys.info()["user"] != "alabadi"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

```{r, eval=FALSE}
## link data folder
# system("ln -s /Users/alabadi/Projects/analysis/R/multiOmics/Ricard/scNMT_gastrulation/data /Users/alabadi/Projects/analysis/R/multiOmics/scNMT_seq_gatrulation")
```

## I/O

```{r, eval=TRUE}
#########
## I/O ##
#########
io$data <- io$out <- io <- list()
## Rproject/slurm file home - must be FULL PATH so the extracted R code can run as well
io$home <- ifelse(params$on_mac,
       "..",
       "/data/cephfs/punim0613/AL/Projects/analysis/R/multiOmics/scNMT_seq_gatrulation")
## outputs
io$outdir <- file.path(io$home, "output")
```

## utils

```{r, eval=TRUE}
io$home
## ----------- source utils ----------- 
## list all R files in utils folder recursively
utls <- list.files(file.path(io$home, "src/utils"), pattern = ".R", full.names = TRUE, recursive = TRUE)
## source all
invisible(sapply(utls, source))
```

```{r, eval=on_hpc}
installer(pkgs = c("data.table", "magrittr", "mixOmics", "scater", "purrr", "uwot"))
```

# Packages

```{r, eval=TRUE}
setup_libPaths()

suppressMessages(library(data.table))
suppressMessages(library(scater))
suppressMessages(library(purrr))
suppressMessages(library(magrittr))
suppressMessages(library(mixOmics))
```


# Initialise

## options

```{r, eval=TRUE}
#############
## Options ##
#############
opts <- list()
rtimes <- list() ## record run times
# Multiple testing correction options
# opts$threshold_fdr  <- 0.10

opts$met.annos <- c(
  Promoters = "prom_2000_2000",
  Genebody = "genebody",
  CGI = "CGI",
  p300 = "ESC_p300",
  CTCF = "ESC_CTCF",  ##CTCF binding sites.  The primary role of CTCF transcription factors is thought to be in regulating the 3D structure of chromatin.
  DHS = "ESC_DHS" ## In these specific regions of the genome, chromatin has lost its condensed structure, exposing the DNA and making it accessible.
  
)

opts$acc.annos <- opts$met.annos
```

## data directories

```{r, eval=TRUE}
## files/directories from data-alias that we'll need
io$data$base   <- file.path(io$home, "data/data-alias/gastrulation") ## data dir
io$data$sample.metadata <- file.path(io$data$base,"sample_metadata.txt")
# io$annos.dir  <- file.path(io$base_dir, "features/filt")
io$data$rna.file   <- file.path(io$data$base, "rna/parsed/SingleCellExperiment.rds")
io$data$met.dir   <- file.path(io$data$base, "met/parsed")
io$data$acc.dir   <- file.path(io$data$base, "acc/parsed")
```

## data files

```{r, eval=TRUE}
## input files
io$data$mets <- lapply(opts$met.annos, function(x) sprintf("%s/%s.tsv.gz", io$data$met.dir, x))
io$data$accs <- lapply(opts$acc.annos, function(x) sprintf("%s/%s.tsv.gz", io$data$acc.dir, x))
```

```{r, eval=on_mac}
## check that all files/dirs are valid
check_files_exist(io)
```

## copy data

```{r}
rapply(io$data, function(x) file.exists(x))
```

```{r, eval=on_mac}
## symlink only the needed datasets from full alias dir to ./data
## so only them are rsynced to hpc
rapply(io$data, function(x) cp(x, pattern = "/data/data-alias/", repl = "/data/" ))
```


```{r, eval=TRUE}
## now replace all /data-alias/ with /data/ in io$data directories
io$data <- rapply(io$data, function(x) sub(pattern = "/data/data-alias/", replacement = "/data/", x = x), how = "list")
```

```{r, eval=TRUE}
## check that all files/dirs are valid
check_files_exist(io)
```

# Data

## metadata

```{r, eval=TRUE}
sample_metadata <- fread(io$data$sample.metadata)
# sample_metadata %>% head() %>% View()
sample_metadata <- sample_metadata %>% 
  .[pass_metQC==TRUE & pass_rnaQC==TRUE  & pass_accQC==TRUE ] %>% 
  .[,c("sample","id_met","id_acc","id_rna","stage","lineage10x_2")] %>%
  .[,stage_lineage:=as.factor(paste(stage,lineage10x_2,sep="_"))]

length(unique(sample_metadata$sample)) ## sample size
```

```{r}
# Define which stage and lineages to look at 
# opts$stages <- c("E5.5","E6.5","E7.5")
# opts$lineages <- c(
#   "Nascent mesoderm",
#   "Primitive Streak",
#   "Mixed mesoderm",
#   "Epiblast",
#   "Rostral neurectoderm",
#   "Gut",
#   "Embryonic endoderm",
#   "Paraxial mesoderm",
#   "Notochord",
#   "Pharyngeal mesoderm",
#   "Mature mesoderm",
#   "Intermediate mesoderm",
#   "Somitic mesoderm",
#   "Caudal mesoderm",
#   "Rostral neuroectoderm",
#   "Caudal neurectoderm",
#   "Caudal Mesoderm",
#   "Def. endoderm"
# )
# opts$stage_lineage <- as.vector(outer(opts$stages,opts$lineages, paste, sep="_"))
# opts$stage_lineage <- opts$stage_lineage[!opts$stage_lineage %in% c("E7.5_Epiblast","E7.5_Primitive Streak")]

# # Define which cells to use
# tmp <- fread(io$data$sample.metadata) %>% 
#   .[,stage_lineage:=paste(stage,lineage10x,sep="_")] %>%
#   # .[assay=="scNMT"] %>%
#   .[stage_lineage%in%opts$stage_lineage]
# opts$met_cells <- tmp %>% .[pass_metQC==T,id_met]
# opts$rna_cells <- tmp %>% .[pass_rnaQC==T,id_rna]
# opts$acc_cells <- tmp %>% .[pass_accQC==T,id_acc]
```


## expression data

### load

<!-- Load RNA data -->
```{r, eval=TRUE}
# Load scater object
sce <- readRDS(io$data$rna.file)

# rna_passQC <- sample_metadata[pass_rnaQC==TRUE]$id_rna
# Filter cells
sce <- sce[,colnames(sce) %in% sample_metadata$id_rna]
```

### change sample names

```{r, eval=TRUE}
## rename id_rna for sce
id_map <- sample_metadata[,c("sample", "id_rna")] %>% data.frame(row.names = 2)
colnames(sce) <- id_map[colnames(sce),]
```

### gene library sizes

```{r}
## mean gene library sizes
hist(rowMeans(logcounts(sce))[rowMeans(logcounts(sce))>0], breaks = 100)
## mean gene library sizes > exp(0.3)
# hist(rowMeans(logcounts(sce))[rowMeans(counts(sce))>0.3], breaks = 100)
```

### mean-variance

```{r}
## Mean vs variance plot
foo <- data.frame(sd=apply(exprs(sce),1,sd), mean=apply(exprs(sce),1,mean))
ggplot(foo, aes(x=mean, y=sd)) +
  geom_point() +
  stat_smooth() +
  scale_color_manual(values=c("black","red")) +
  xlab('Mean') + ylab('Standard deviation')

```

### pca

```{r, eval=TRUE}
io$out$pca_scnmt <- file.path(io$outdir, "pca_scnmt.rds")
```

```{r, eval=on_hpc}
pca_scnmt <- pca(t(logcounts(sce)), ncomp=10)
saveRDS(pca_scnmt, file = io$out$pca_scnmt)
```

```{r}
pca_scnmt <- readRDS(io$out$pca_scnmt)
# visualise rna
plot(pca_scnmt)

plotIndiv(pca_scnmt, group = sce$stage, pch=16, legend = TRUE)
## comp 2,3
plotIndiv(pca_scnmt, comp = c(2,3), group = sce$stage, pch=16, legend = TRUE)

plotIndiv(pca_scnmt, comp = c(1,3), group = sce$stage, pch=16, legend = TRUE)
```

### umap

```{r, eval=TRUE}
io$out$umap_scnmt <- file.path(io$outdir, "umap_scnmt.rds")
```

```{r, eval=on_hpc}
library(uwot)
Ncomps = 3
umap_scnmt <- uwot::umap(t(logcounts(sce)), n_components = Ncomps) %>%
  as.data.frame() %>%
  set_colnames(paste0("UMAP_", seq_len(Ncomps)))
saveRDS(umap_scnmt, file = io$out$umap_scnmt )
```

```{r}
umap_scnmt <- readRDS(io$out$umap_scnmt)
library(ggplot2)
ggplot(umap_scnmt) + geom_point(aes(UMAP_1, UMAP_2, col = sce$stage))
ggplot(as.data.frame(umap_scnmt)) + geom_point(aes(UMAP_1, UMAP_3, col = sce$stage))
```


## epigenetic data

### process

#### methylation
```{r}
## function to read data and filter
# io <- list()
# io$data$met.dir <- "../data/gastrulation/met/parsed"
# io$data$acc.dir <- "../data/gastrulation/acc/parsed"
```

```{r, eval=TRUE}
io$out$met_dt_list <- file.path(io$outdir, "met_dt_list.rds")
```

##### calculate site stats

```{r, eval=on_hpc}
rtimes$met_dt_list <- system.time({
  
  met_dt_list <- mclapply(io$data$mets, function(tsv_name) {
    
    calc_site_stats(filePath = tsv_name,
                    sample_name = "id_met", ## name of sample to output
                    keep_samples = sample_metadata$id_met, ## pass QC samples
                    min_N = 3, ## min number of calls at site
                    min_cov = 400, ## min number of cells detecting
                    alpha = 0.1) ## for lower bound of CI
  }, mc.cores = parallel::detectCores())
  
})

saveRDS(met_dt_list, file = io$out$met_dt_list)
```

```{r, eval=TRUE}
io$out$met_dist_mats <- file.path(io$outdir, "met_dist_mats.rds")
```

##### get distance matrices

```{r, eval=FALSE}
# met_dist_mats <- list()
# 
# for (dt_name in names(met_dt_list)) {
#   met_dist_mats[[dt_name]] <- wt_euc_dist(met_dt = met_dt_list[[dt_name]], 
#                                           sample_name = "id_met", 
#                                           n_hvr = 2000)
#   gc()
# }
# 
# saveRDS(met_dist_mats, file = io$out$met_dist_mats)
```

```{r, eval=on_hpc}
n_hvr <- 1000
for (dt_name in names(met_dt_list)) {
  dist_mat <- wt_euc_dist(met_dt = met_dt_list[[dt_name]], 
                                          sample_name = "id_met", 
                                          n_hvr = n_hvr)
  saveRDS(dist_mat, file = sprintf("%s/met_dist_mat_%s_%s.rds", io$outdir, dt_name, n_hvr))
  rm(dist_mat)
  gc()
}
```


#### accessibility

```{r, eval=TRUE}
io$out$acc_dt_list <- file.path(io$outdir, "acc_dt_list.rds")
```

##### calculate site stats

```{r, eval=on_hpc}
rtimes$acc_dt_list <- system.time({
  
  acc_dt_list <- lapply(io$data$accs, function(tsv_name) {
    
    calc_site_stats(filePath = tsv_name,
                    sample_name = "id_acc", ## name of sample to output
                    keep_samples = sample_metadata$id_acc, ## pass QC samples
                    min_N = 3, ## min number of calls at site
                    min_cov = 500, ## min number of cells detecting
                    alpha = 0.1) ## for lower bound of CI
  })
  
})

saveRDS(acc_dt_list, file = io$out$acc_dt_list)
```

```{r, eval=TRUE}
io$out$acc_dist_mats <- file.path(io$outdir, "acc_dist_mats.rds")
```

##### get distance matrices

```{r, eval=FALSE}
acc_dist_mats <- list()

for (dt_name in names(acc_dt_list)) {
  acc_dist_mats[[dt_name]] <- wt_euc_dist(met_dt = acc_dt_list[[dt_name]], 
                                          sample_name = "id_acc", 
                                          n_hvr = 2000)
  gc()
}

saveRDS(acc_dist_mats, file = io$out$acc_dist_mats)
```

```{r, eval=on_hpc}
n_hvr <- 1000
for (dt_name in names(acc_dt_list)) {
  dist_mat <- wt_euc_dist(met_dt = acc_dt_list[[dt_name]], 
                                          sample_name = "id_acc", 
                                          n_hvr = n_hvr)
  saveRDS(dist_mat, file = sprintf("%s/acc_dist_mat_%s_%s.rds", io$outdir, dt_name, n_hvr))
  rm(dist_mat)
  gc()
}
```

### post-process

#### methylation

```{r}
met_dist_mats <- readRDS(io$out$met_dist_mats)
```

##### MDS

```{r}
cmdscale(d = met_dist_mats[[1]]) %>% heatmap()
```

##### dcast

create wide matrices for MAE

```{r}
rtimes$dcast_met <- system.time({
  dcast_met <- dcast_dt_list(met_dt_list, row = "id", col = "id_met", value.var = "rate")
}) ## 8 secs

```

```{r}
lapply(dcast_met, dim)
```


```{r}
## quick check datasets to ensure features are properly ranked
mapply(x=dcast_met, y=met_dt_list, FUN = function(x,y) {
  ## is the first rowname in matrices the same as the max lci in data.table?
  identical(x %>% rownames() %>% .[1],
            y %>% .[which.max(.$lci)] %>% .$id)
})
```

#### accessibility

```{r}
acc_dist_accs <- readRDS(io$out$acc_dist_mats)
```

##### MDS

```{r}
cmdscale(d = acc_dist_mats[[1]]) %>% heatmap()
```

##### dcast

create wide matrices for MAE

```{r}
rtimes$dcast_acc <- system.time({
  dcast_acc <- dcast_dt_list(acc_dt_list, row = "id", col = "id_acc", value.var = "rate")
})

```

```{r}
lapply(dcast_acc, dim)
```

```{r}
## quick check datasets to ensure features are properly ranked
mapply(x=dcast_acc, y=acc_dt_list, FUN = function(x,y) {
  ## is the first rowname in matrices the same as the max lci in data.table?
  identical(x %>% rownames() %>% .[1],
            y %>% .[which.max(.$lci)] %>% .$id)
})
```


# MultiAssayExperiment

```{r}
library(MultiAssayExperiment)

exper.list <- ExperimentList(rna = logcounts(sce),
                           
                           met_genebody = dcast_met$Genebody,
                           met_promoter = dcast_met$Promoters,
                           met_cgi = dcast_met$CGI,
                           met_p300 = dcast_met$p300,
                           met_CTCF = dcast_met$CTCF,
                           met_DHS = dcast_met$DHS,
                           
                           acc_genebody = dcast_acc$Genebody,
                           acc_promoter = dcast_acc$Promoters,
                           acc_cgi = dcast_acc$CGI,
                           acc_p300 = dcast_acc$p300,
                           acc_CTCF = dcast_acc$CTCF,
                           acc_DHS = dcast_acc$DHS
                           )
```


```{r}
coldata <- sample_metadata %>% data.frame(row.names = TRUE) %>% .[,c("stage", "lineage10x_2", 
"stage_lineage")] %>% DataFrame()
scnmtseq_gastrulation_mae <- MultiAssayExperiment(experiments = exper.list, colData = coldata)
```


```{r}
saveRDS(scnmtseq_gastrulation_mae, file= "../output/scnmtseq_gastrulation_mae_826-cells_orderedFeatures.rds")
```

<!-- Add decomposed (biological) variance to rna for HVGS selection: -->

```{r}
# scnmtseq_gastrulation_mae <- readRDS("../output/scnmtseq_gastrulation_mae.rds")
```

```{r}
# sce <- readRDS("../data/gastrulation/rna/parsed/SingleCellExperiment.rds")
# rowData(sce)
```

```{r}
library(scran)
fit <- trendVar(sce, use.spikes=FALSE)
decomp.var <- decomposeVar(sce, fit = fit)
library(ggplot2)
ggplot(as.data.frame(decomp.var)) + geom_point(aes(x=mean, y=total), alpha=0.6) + theme_bw()
ggplot(as.data.frame(decomp.var)) + geom_point(aes(x=mean, y=bio), alpha=0.6) + theme_bw()
```

```{r}
## add to rowData
rowData(sce) <- cbind(rowData(sce), decomp.var[,c("bio", "total")])
```

```{r}
library(MultiAssayExperiment)

exper.list <- ExperimentList(rna = sce,
                           experiments(scnmtseq_gastrulation_mae)[-1]
                           )

MultiAssayExperiment(experiments = exper.list, colData = colData(scnmtseq_gastrulation_mae))

coldata <- sample_metadata %>% data.frame(row.names = TRUE) %>% .[,c("stage", "lineage10x_2", 
"stage_lineage")] %>% DataFrame()
scnmtseq_gastrulation_mae <- MultiAssayExperiment(experiments = exper.list, colData = coldata)
scnmtseq_gastrulation_mae <- MatchedAssayExperiment(mae_scnmt_seq)

coldata <- sample_metadata %>% data.frame(row.names = TRUE) %>% .[,c("stage", "lineage10x_2", 
"stage_lineage")] %>% DataFrame()
scnmtseq_gastrulation_mae <- MultiAssayExperiment(experiments = exper.list, colData = coldata)
scnmtseq_gastrulation_mae <- MatchedAssayExperiment(mae_scnmt_seq)
```


